libbpf-tools/syscount.c

```diff
@@ -0,0 +1,1167 @@
+	[281] = "epoll_pwait",
+	[282] = "signalfd",
+	[283] = "timerfd_create",
+	[284] = "eventfd",
+	[285] = "fallocate",
+	[286] = "timerfd_settime",
+	[287] = "timerfd_gettime",
+	[288] = "accept4",
+	[289] = "signalfd4",
+	[290] = "eventfd2",
+	[291] = "epoll_create1",
+	[292] = "dup3",
+	[293] = "pipe2",
+	[294] = "inotify_init1",
+	[295] = "preadv",
+	[296] = "pwritev",
+	[297] = "rt_tgsigqueueinfo",
+	[298] = "perf_event_open",
+	[299] = "recvmmsg",
+	[300] = "fanotify_init",
+	[301] = "fanotify_mark",
+	[302] = "prlimit64",
+	[303] = "name_to_handle_at",
+	[304] = "open_by_handle_at",
+	[305] = "clock_adjtime",
+	[306] = "syncfs",
+	[307] = "sendmmsg",
+	[308] = "setns",
+	[309] = "getcpu",
+	[310] = "process_vm_readv",
+	[311] = "process_vm_writev",
+	[312] = "kcmp",
+	[313] = "finit_module",
+	[314] = "sched_setattr",
+	[315] = "sched_getattr",
+	[316] = "renameat2",
+	[317] = "seccomp",
+	[318] = "getrandom",
+	[319] = "memfd_create",
+	[320] = "kexec_file_load",
+	[321] = "bpf",
+	[322] = "execveat",
+	[323] = "userfaultfd",
+	[324] = "membarrier",
+	[325] = "mlock2",
+	[326] = "copy_file_range",
+	[327] = "preadv2",
+	[328] = "pwritev2",
+	[329] = "pkey_mprotect",
+	[330] = "pkey_alloc",
+	[331] = "pkey_free",
+	[332] = "statx",
+	[333] = "io_pgetevents",
+	[334] = "rseq",
+	[424] = "pidfd_send_signal",
+	[425] = "io_uring_setup",
+	[426] = "io_uring_enter",
+	[427] = "io_uring_register",
+	[428] = "open_tree",
+	[429] = "move_mount",
+	[430] = "fsopen",
+	[431] = "fsconfig",
+	[432] = "fsmount",
+	[433] = "fspick",
+	[434] = "pidfd_open",
+	[435] = "clone3",
+	[437] = "openat2",
+	[438] = "pidfd_getfd",
+};
+size_t syscall_names_x86_64_size = sizeof(syscall_names_x86_64)/sizeof(char*);
+#endif
+
+static void syscall_name(unsigned n, char *buf, size_t size)
+{
+	const char *name = NULL;
+
+	if (n < syscall_names_size)
+		name = syscall_names[n];
+#ifdef __x86_64__
+	else if (n < syscall_names_x86_64_size)
+		name = syscall_names_x86_64[n];
+#endif
+
+	if (name)
+		strncpy(buf, name, size-1);
+	else
+		snprintf(buf, size, "[unknown: %u]", n);
+}
+
+static int list_syscalls(void)
+{
+	const char **list = syscall_names;
+	size_t size = syscall_names_size;
+
+#ifdef __x86_64__
+	if (!size) {
+		size = syscall_names_x86_64_size;
+		list = syscall_names_x86_64;
+	}
+#endif
+
+	for (size_t i = 0; i < size; i++) {
+		if (list[i])
+			printf("%3zd: %s\n", i, list[i]);
+	}
+
+	return (!list || !size);
+}
+
+static int compar_count(const void *dx, const void *dy)
+{
+	__u64 x = ((struct data_t *) dx)->count;
+	__u64 y = ((struct data_t *) dy)->count;
+	return x > y ? -1 : !(x == y);
+}
+
+static int compar_latency(const void *dx, const void *dy)
+{
+	__u64 x = ((struct data_t *) dx)->total_ns;
+	__u64 y = ((struct data_t *) dy)->total_ns;
+	return x > y ? -1 : !(x == y);
+}
+
+static const char *comm(pid_t pid, char *buf, size_t size)
+{
+	char path[32];
+	int fd;
+	int n;
+
+	snprintf(path, sizeof(path), "/proc/%d/comm", pid);
```

Comment:
this pid to comm mapping from procfs is racy, process could be dead or PID can even be reused in extreme cases. BPF code can easily capture this, so why not do that?

Reply:
ping? this approach doesn't work for short-lived processes at all. You can have a separate array for this or just extend start and store process name first time. Is there anything wrong with that?